### PR TITLE
Regular attention flash decoding refactor and bug fixes

### PR DESCRIPTION
**Pull request by bdevorem**
## Motivation

Flash decoding in `fuse_attention` rebuilds a fused attention submodule where shapes have been changed to support parallelizable split sequence length dimensions. This requires semantically equivalent updates to any ops that work on specific dimensions, including reductions and broadcasts. That rebuild was brittle on real transformer graphs and could fail or produce incorrect IR when the attention block included:
- **`@literal` constants** fused into the submodule (e.g. BERT `1/sqrt(d)` scale)
- **Extra `@param` inputs** with the attention-score shape (e.g. attention masks)
- **`@outline` nodes** inside the submodule
- **Unary ops between softmax broadcasts and consumers** (e.g. `contiguous` between `multibroadcast(reduce_max)` and `sub`)

The old rebuild logic assumed only Q/K/V params and inferred `multibroadcast` target shapes via a sibling-input hack on `sub`/`div` parents. That broke on BERT-like patterns and other common attention variants.

Fixes [AIMIGRAPHX-1005], [AIMIGRAPHX-1227], [AIMIGRAPHX-1228].

Additionally, the previous implementation of flash decoding did not handle causal masks that needed to be both padded and broadcasted to a new split shape. Supporting that is a non-trivial amount of code, so support for uneven KV sequence splitting is removed for now; a follow-up PR will add padding support for uneven splits.

Note: tests, description, function boilerplate, additional param support made with the help of cursor

## Technical Details

- Handle **`@literal`** and **`@outline`** during rebuild
- Add **`transform_score_literal`** to split broadcastable mask literals at compile time on their **own** layout (`{1,1,M,N}` -> `{1,1,G,M,N/G}`), so `multibroadcast` still covers the leading `1`s and constants are not expanded to the full `[B,H,G,M,N/G]` tensor. Broadcast-only constants (scale, `-inf`) keep their original shape.
- Replace the sibling-input `multibroadcast` hack. Drive `multibroadcast` rewriting from the gemm1 score shape instead of walking consumers to find a broadcast target.
- `multibroadcast` rebuild fallback: if an input cannot broadcast directly to the split shape but can broadcast to the original score shape, broadcast first and then split into the split layout.
- `can_multibroadcast` helper in `src/common.cpp` / `common.hpp`
- Fix score-shaped splits. `[B,H,M,N]` -> `[B,H,G,M,N/G]` requires a reshape to `[B,H,M,G,N/G]` followed by a transpose moving `G` ahead of `M`; a plain reshape interleaves `M` and `N` and produces wrong data. Split scores come from `Q@K'` where `K` was split as `n = g*(N/G) + j`, so `split[..., g, m, j]` must equal `in[..., m, g*(N/G) + j]`. Applies to score-shaped `@param` inputs and to the `multibroadcast` fallback above.
- Rebuild the submodule with `module::add_instructions` and an `inserter` that performs the shape/rank rewrites, instead of a hand-rolled instruction walk. `@param`, `@outline` and `@return` are handled by the framework; literals are split and mapped up front, since `add_instructions` copies them verbatim without consulting the inserter.
- Skip flash decoding up front for ops that cannot be rewritten, before any instructions are built: score-shaped `broadcast` (its `axis` attribute would need adjusting as the rank grows) and reductions that are not over a single innermost axis. These were previously a mid-rebuild abort and asserts.
- Deterministic IR. Insert the input reshapes in submodule-parameter order rather than iterating an `unordered_map` while modifying the module.
- `transform_score_literal` builds the split as a strided view of the literal's own buffer instead of per-element index arithmetic.
- Express the Q/K/V splits as op lists on `transformed_shapes_result`, applied with a shared `insert_ops` fold, so each transform is described in one place.

### Tests (`test/fuse_attention.cpp`, `test/shape_test.cpp`)
- Helpers `flash_decoding_submodule` / `flash_decoding_has` for asserting on the rebuilt submodule.
- New coverage for:
  - BERT-like attention with an embedded scale `@literal`
  - 4D attention with a mask `@param` and `@literal` constants
  - Broadcastable mask `@literal` split at compile time on its own layout
  - Broadcastable mask `@param` using the broadcast-then-split fallback
  - Score-shaped `@param` split preserves values (`split[..., g, m, j] == in[..., m, g*(N/G) + j]`)
  - Unary ops on softmax broadcast paths
  - Submodules whose instruction order places a broadcast before its consumer's gemm
  - `@outline` nodes in submodules
  - LSE attention correctly skipped
  - FP32 softmax intermediate (`convert` around softmax)
  - Existing 3D flash-decoding cases (rectangular, auto-split)
- Uneven-split padding test is skipped, matching the removed padding support.
- Add `can_multibroadcast` unit tests in `test/shape_test.cpp`.

## Changelog Category

Add a `CHANGELOG.md` entry for any option other than `Not Applicable`
- - [ ] Added: New functionality.
- - [x] Changed: Changes to existing functionality.
- - [ ] Removed: Functionality or support that has been removed. (Compared to a previous release)
- - [ ] Optimized: Component performance that has been optimized or improved.
- - [x] Resolved Issues: Known issues from a previous version that have been resolved.
- - [ ] Not Applicable: This PR is not to be included in the changelog.

Follow the [LLVM AI Tool Use Policy](https://llvm.org/docs/AIToolPolicy.html) for contributions using AI.

[AIMIGRAPHX-1005]: https://amd-hub.atlassian.net/browse/AIMIGRAPHX-1005?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/fuse_attention.cpp [hunk 1]
```diff
@@ -201,8 +203,8 @@ struct find_transposed_attention
                 return false;
             return std::equal(
                        perm.begin(), perm.end() - 2, migraphx::range(perm.size() - 2).begin()) and
-                   perm[perm.size() - 2] == static_cast<int64_t>(perm.size() - 1) and
-                   perm[perm.size() - 1] == static_cast<int64_t>(perm.size() - 2);
+                   perm[perm.size() - 2] == perm.size() - 1 and
+                   perm[perm.size() - 1] == perm.size() - 2;
```
[the remark this answers is not in the document]
Comment (bdevorem) [reply src/fuse_attention.cpp]: reverted

File: test/fuse_attention.cpp [hunk 2]
```diff
@@ -43,12 +44,68 @@
...
 {
     migraphx::run_passes(p, {fa, migraphx::dead_code_elimination{}});
 }
 
+// Run fuse_attention without inheriting flash-decoding settings from the environment.
+static void run_pass_isolated(migraphx::program& p, migraphx::fuse_attention fa = {})
+{
+    const char* splits_env    = std::getenv("MIGRAPHX_FLASH_DECODING_NUM_SPLITS");
+    const char* enabled_env   = std::getenv("MIGRAPHX_FLASH_DECODING_ENABLED");
+    const char* threshold_env = std::getenv("MIGRAPHX_FLASH_DECODING_THRESHOLD");
```
[the remark this answers is not in the document]
Comment (bdevorem) [reply test/fuse_attention.cpp]: tests are removed, cursor added them accidentally

File: src/fuse_attention.cpp [hunk 3]
```diff
@@ -534,14 +536,227 @@ struct find_flash_decoding
         return result;
     }
 
+    static std::vector<size_t> get_scores_split_lens(const std::vector<size_t>& lens,
+                                                     std::size_t num_groups)
+    {
+        assert(lens.size() >= 2);
+        assert(lens.back() % num_groups == 0);
```
[the remark this answers is not in the document]
Comment (bdevorem) [reply src/fuse_attention.cpp:538]: done

File: src/fuse_attention.cpp [hunk 4]
```diff
@@ -501,132 +510,240 @@ struct find_flash_decoding
...
 
         // V: [B, N, D] -> [B, G, N/G, D] via direct reshape
         result.v_shape                            = insert_g(v_lens);
         result.v_shape[result.v_shape.size() - 2] = n_split;
+        result.v_ops = {make_op("reshape", {{"dims", result.v_shape}})};
+
+        return result;
+    }
+
+    // Lens of a score-shaped tensor after its key axis is split:
```
Comment (pfultz2) [inline src/fuse_attention.cpp:531]: What is a "score-shaped tensor"? Is this the tensor where we add the extra split dim?

**Comment by pfultz2** [comment]
> Additionally, the previous implementation of flash decoding did not handle causal masks that needed to be both padded and broadcasted to a new split shape. Supporting that is a non-trivial amount of code, so support for uneven KV sequence splitting is removed for now; a follow-up PR will add padding support for uneven splits.

Uneven splits are unlikely, and split_dim finds the number of splits that are a multiple of N so we can hold off of this.

Overall the approach looks good.

**Comment by TedThemistokleous** [comment]
> > Additionally, the previous implementation of flash decoding did not handle causal masks that needed to be both padded and broadcasted to a new split shape. Supporting that is a non-trivial amount of code, so support for uneven KV sequence splitting is removed for now; a follow-up PR will add padding support for uneven splits.
> 
> Uneven splits are unlikely, and split_dim finds the number of splits that are a multiple of N so we can hold off of this.
> 
> Overall the approach looks good.

I think this looks good - About to approve - do we still want this?

**Comment by pfultz2** [comment]
> > > Additionally, the previous implementation of flash decoding did not handle causal masks that needed to be both padded and broadcasted to a new split shape. Supporting that is a non-trivial amount of code, so support for uneven KV sequence splitting is removed for now; a follow-up PR will add padding support for uneven splits.
> > 
> > 
> > Uneven splits are unlikely, and split_dim finds the number of splits that are a multiple of N so we can hold off of this.
> > Overall the approach looks good.
> 
> I think this looks good - About to approve - do we still want this?

Yes I think so.

**Comment by bdevorem** [comment]
> > > Additionally, the previous implementation of flash decoding did not handle causal masks that needed to be both padded and broadcasted to a new split shape. Supporting that is a non-trivial amount of code, so support for uneven KV sequence splitting is removed for now; a follow-up PR will add padding support for uneven splits.
> > 
> > 
> > Uneven splits are unlikely, and split_dim finds the number of splits that are a multiple of N so we can hold off of this.
> > Overall the approach looks good.
> 
> I think this looks good - About to approve - do we still want this?

Yes. I'll fix the merge conflicts and then I could use both of your approvals @TedThemistokleous @pfultz2

**Review by TedThemistokleous — APPROVED** [review]
